src/Common/FileCache.cpp

```diff
@@ -473,40 +641,42 @@ bool LRUFileCache::tryReserve(const Key & key, size_t offset, size_t size, std::
     else
     {
         size_t removed_size = 0;
-        size_t queue_size = queue.getElementsNum(cache_lock);
+        size_t queue_size = main_priority->getElementsNum(cache_lock);
 
         auto * cell_for_reserve = getCell(key, offset, cache_lock);
 
-        std::vector<IFileCache::LRUQueue::Iterator> ghost;
+        std::vector<std::tuple<Key, size_t, size_t>> ghost;
         std::vector<FileSegmentCell *> trash;
         std::vector<FileSegmentCell *> to_evict;
 
         auto is_overflow = [&]
         {
-            return (max_size != 0 && queue.getTotalCacheSize(cache_lock) + size - removed_size > max_size)
+            return (max_size != 0 && main_priority->getCacheSize(cache_lock) + size - removed_size > max_size)
             || (max_element_size != 0 && queue_size > max_element_size)
             || (query_context->getCacheSize() + size - removed_size > query_context->getMaxCacheSize());
         };
 
         /// Select the cache from the LRU queue held by query for expulsion.
-        for (auto iter = query_context->queue().begin(); iter != query_context->queue().end(); iter++)
+        for (auto iter = query_context->getPriority()->getLowestPriorityWriteIterator(cache_lock); iter->valid();)
         {
             if (!is_overflow())
                 break;
 
-            auto * cell = getCell(iter->key, iter->offset, cache_lock);
+            auto * cell = getCell(iter->key(), iter->offset(), cache_lock);
 
             if (!cell)
             {
                 /// The cache corresponding to this record may be swapped out by
                 /// other queries, so it has become invalid.
-                ghost.push_back(iter);
-                removed_size += iter->size;
+                removed_size += iter->size();
+                ghost.push_back({iter->key(), iter->offset(), iter->size()});
+                /// next()
+                iter->remove(cache_lock);
```

Comment:
May be rename this method to `removeAndGetNext()` so it will be more obvious that the iterator will be incremented.
btw may be make the behaviour more consistent with canonical iterators? so `remove` method will not increment the iterator itself, but return iterator pointing to the next position after currently deleted. Then you can just reassign it to `iter`. What do you think?